pkg/cpuallocator/allocator.go

```diff
@@ -389,6 +402,526 @@ func (a *allocatorHelper) takeIdleClusters() {
+			total := full + part
+
+			if diePartA >= total && diePartB < total {
+				return -1
+			}
+			if diePartA < total && diePartB >= total {
+				return 1
+			}
+			if diePartA >= total && diePartB >= total {
+				if diff := diePartA - diePartB; diff != 0 {
+					return diff
+				}
+				// for a tie prefer looser (bigger) group, smaller package, die, and group IDs
+				if diff := csetA.Size() - csetB.Size(); diff != 0 {
+					return -diff
+				}
+				if gA.pkg != gB.pkg {
+					return gA.pkg - gB.pkg
+				}
+				if gA.die != gB.die {
+					return gA.die - gB.die
+				}
+				return gA.id - gB.id
+			}
+
+			// equality: both are unusable, don't need to sort them
+			return 0
+		}
+
+		sorter = &cacheGroupSorter{
+			pick:       pickGroups,
+			sortPrefer: sortIdle,
+			sortUsable: sortUsed,
+		}
+	)
+
+	log.Debug("looking for %d CPUs (prio %s) from %s", a.cnt, a.prefer, a.from)
+
+	sorter.sortCacheGroups(a)
+
+	var (
+		preferPkgCPUs int
+		usablePkgCPUs int
+		chosenPkg     int
+
+		result = a.result
+		from   = a.from
+		cnt    = a.cnt
+	)
+
+	switch {
+	case len(sorter.prefer) > 0:
+		chosenPkg = sorter.prefer[0].pkg
+		preferPkgCPUs = sorter.preferPkgCPUCount(chosenPkg)
+		usablePkgCPUs = sorter.usablePkgCPUCount(chosenPkg)
+	case len(sorter.usable) > 0:
+		chosenPkg = sorter.usable[0].pkg
+		usablePkgCPUs = sorter.usablePkgCPUCount(chosenPkg)
+	}
+
+	if preferPkgCPUs+usablePkgCPUs < a.cnt {
+		log.Debug("=> no package can satisfy the allocation")
+		return
+	}
+
+	//
+	// take full idle cache groups, splitting up the last one if necessary
+	//
+
+	log.Debug("trying to take idle cache groups...")
+	for i, g := range sorter.prefer {
+		if cnt == 0 {
+			break
+		}
+
+		cset := sorter.cpus[g]
+
+		if cnt >= cset.Size() {
+			log.Debug("=> took full idle cache group %d. %s", i, g)
+
+			result = result.Union(cset)
+			from = from.Difference(cset)
+			cnt -= cset.Size()
+			continue
+		}
+
+		// partially allocate the rest from this group
+		ta := newAllocatorHelper(a.sys, a.topology)
+		ta.prefer = a.prefer
+		ta.flags = AllocIdleCores
+		ta.from = cset
+		ta.cnt = cnt
+		use := ta.allocate()
+
+		log.Debug("=> took %d CPUs (%s) from idle cache group %d. %s", use.Size(), use, i, g)
+
+		result = result.Union(use)
+		from = from.Difference(use)
+		cnt -= use.Size()
+	}
+
+	if cnt == 0 {
+		a.result = result
+		a.from = from
+		a.cnt = cnt
+		return
+	}
+
+	//
+	// allocate non-idle usable cache groups
+	//
+	// We try a few strategies to fulfill the allocation in this order:
+	//   1. try to find a single group with the exact number of CPUs
+	//   2. try to find the minimal number of same-sized groups
+	//   3. fulfill request by taking groups in decreasing size order
+	//
+
+	log.Debug("%d more CPUs needed", cnt)
+
+	var (
+		groupsBySize = map[int][]*cacheGroup{}
+		totalByIndex = make([]int, 0, len(sorter.usable))
+		totalCPUs    = 0
+	)
+
+	for i := 0; i <= len(sorter.usable)-1; i++ {
+		g := sorter.usable[i]
+		cset := sorter.cpus[g]
```

Comment:
Question: is it possible that due to CPU offlining or fully used cache groups or something else it could be that cset.Size() == 0?

If so, maybe these groups should be skipped over here to avoid the risk of division-by-zero on line 828.

Reply:
But I can add a `grpSize > 0` safeguard check there...

Reply:
Added.